File: apps/web/src/app/books/[id]/read/reader-types.ts
import { AudioSyncSegment, AudioSyncStatus, ReaderChapter } from '@psc/shared-types';

/** One continuously-scrolling block in the reader — a Topic, or one of its Subtopics. */
export interface ReadingUnit {
  unitIndex: number;
  chapterId: string;
  chapterTitle: string;
  /** 1-based position of this unit's chapter, for the "Chapter 3" divider heading. */
  chapterNumber: number;
  /** True on the first unit of each chapter — the reader renders a divider above it. */
  isChapterStart: boolean;
  /** 1-based position of this topic within its chapter (subtopics share their parent's number). */
  topicNumber: number;
  /** The Topic this unit resumes to — ReadingProgress only tracks topic granularity, so a subtopic unit still saves its parent topic's id. */
  topicId: string;
  kind: 'topic' | 'subtopic';
  id: string;
  title: string;
  description?: string | null;
  youtubeUrl?: string | null;
  audioUrl?: string | null;
  pdfUrl?: string | null;
  audioSyncStatus?: AudioSyncStatus | null;
  audioSyncSegments?: AudioSyncSegment[] | null;
}

export interface ChapterSummary {
  chapterId: string;
  chapterNumber: number;
  title: string;
  orderIndex: number;
  unitStart: number;
  unitEnd: number;
  hasYoutube?: boolean;
  hasAudio?: boolean;
  hasPdf?: boolean;
  topics?: {
    id: string;
    title: string;
    unitIndex: number;
    chapterNumber: number;
    topicNumber: number;
    kind: 'topic' | 'subtopic';
    youtubeUrl?: string | null;
    audioUrl?: string | null;
    pdfUrl?: string | null;
  }[];
}

export function flattenChapters(chapters: ReaderChapter[]): ReadingUnit[] {
  const units: ReadingUnit[] = [];
  let chapterNumber = 0;

  for (const chapter of chapters) {
    if (chapter.topics.length === 0) continue;
    chapterNumber += 1;
    let topicNumber = 0;
    let isChapterStart = true;

    for (const topic of chapter.topics) {
      topicNumber += 1;
      units.push({
        unitIndex: units.length,
        chapterId: chapter.id,
        chapterTitle: chapter.title,
        chapterNumber,
        isChapterStart,
        topicNumber,
        topicId: topic.id,
        kind: 'topic',
        id: topic.id,
        title: topic.title,
        description: topic.description,
        youtubeUrl: topic.youtubeUrl,
        audioUrl: topic.audioUrl,
        pdfUrl: topic.pdfUrl,
        audioSyncStatus: topic.audioSyncStatus,
        audioSyncSegments: topic.audioSyncSegments,
      });
      isChapterStart = false;

      for (const subtopic of topic.subtopics) {
        units.push({
          unitIndex: units.length,
          chapterId: chapter.id,
          chapterTitle: chapter.title,
          chapterNumber,
          isChapterStart: false,
          topicNumber,
          topicId: topic.id,
          kind: 'subtopic',
          id: subtopic.id,
          title: subtopic.title,
          description: subtopic.description,
          youtubeUrl: subtopic.youtubeUrl,
          audioUrl: subtopic.audioUrl,
          pdfUrl: subtopic.pdfUrl,
          audioSyncStatus: subtopic.audioSyncStatus,
          audioSyncSegments: subtopic.audioSyncSegments,
        });
      }
    }
  }
  return units;
}

export function buildChapterSummaries(chapters: ReaderChapter[], units: ReadingUnit[]): ChapterSummary[] {
  return chapters
    .filter((c) => units.some((u) => u.chapterId === c.id))
    .map((chapter) => {
      const chapterUnits = units.filter((u) => u.chapterId === chapter.id);
      const indices = units.map((u, idx) => (u.chapterId === chapter.id ? idx : -1)).filter((idx) => idx >= 0);
      const chapterNumber = chapterUnits[0]?.chapterNumber || 1;
      return {
        chapterId: chapter.id,
        chapterNumber,
        title: chapter.title,
        orderIndex: chapter.orderIndex,
        unitStart: indices[0],
        unitEnd: indices[indices.length - 1],
        hasYoutube: chapterUnits.some((u) => Boolean(u.youtubeUrl)),
        hasAudio: chapterUnits.some((u) => Boolean(u.audioUrl)),
        hasPdf: chapterUnits.some((u) => Boolean(u.pdfUrl)),
        topics: chapterUnits.map((u) => ({
          id: u.id,
          title: u.title,
          unitIndex: u.unitIndex,
          chapterNumber: u.chapterNumber,
          topicNumber: u.topicNumber,
          kind: u.kind,
          youtubeUrl: u.youtubeUrl,
          audioUrl: u.audioUrl,
          pdfUrl: u.pdfUrl,
        })),
      };
    });
}

File: mobile/lib/features/books/reader_types.dart
import '../../data/models/book.dart';
import '../../data/models/pdf_sync.dart';

/// One readable block in the reader — a Topic, or one of its Subtopics.
///
/// Ported from the website's `reader-types.ts` so both surfaces number
/// chapters and topics identically and resume to the same place.
class ReadingUnit {
  const ReadingUnit({
    required this.unitIndex,
    required this.chapterId,
    required this.chapterTitle,
    required this.chapterNumber,
    required this.isChapterStart,
    required this.topicNumber,
    required this.topicId,
    required this.isSubtopic,
    required this.id,
    required this.title,
    this.description,
    this.youtubeUrl,
    this.audioUrl,
    this.pdfUrl,
    this.syncCues,
  });

  final int unitIndex;
  final String chapterId;
  final String chapterTitle;

  /// 1-based position of this unit's chapter, for the "Chapter 3" divider.
  final int chapterNumber;

  /// True on the first unit of each chapter.
  final bool isChapterStart;

  /// 1-based position within the chapter; subtopics share their parent's number.
  final int topicNumber;

  /// Progress is only tracked at topic granularity, so a subtopic unit still
  /// saves its parent topic's id.
  final String topicId;

  final bool isSubtopic;
  final String id;
  final String title;
  final String? description;
  final String? youtubeUrl;
  final String? audioUrl;
  final String? pdfUrl;

  /// The unit's PDF↔audio timing map, when one has been authored.
  final PdfSyncMap? syncCues;

  bool get hasAudio => (audioUrl ?? '').isNotEmpty;
  bool get hasVideo => (youtubeUrl ?? '').isNotEmpty;
  bool get hasPdf => (pdfUrl ?? '').isNotEmpty;
  bool get hasMedia => hasAudio || hasVideo || hasPdf;
  bool get hasBody => (description ?? '').trim().isNotEmpty;
}

/// Walks the chapter tree into the flat, ordered list the reader pages through.
/// Chapters with no topics are skipped — there would be nothing to show.
List<ReadingUnit> flattenChapters(List<Chapter> chapters) {
  final units = <ReadingUnit>[];
  var chapterNumber = 0;

  for (final chapter in chapters) {
    if (chapter.topics.isEmpty) continue;
    chapterNumber += 1;
    var topicNumber = 0;
    var isChapterStart = true;

    for (final topic in chapter.topics) {
      topicNumber += 1;
      units.add(
        ReadingUnit(
          unitIndex: units.length,
          chapterId: chapter.id,
          chapterTitle: chapter.title,
          chapterNumber: chapterNumber,
          isChapterStart: isChapterStart,
          topicNumber: topicNumber,
          topicId: topic.id,
          isSubtopic: false,
          id: topic.id,
          title: topic.title,
          description: topic.description,
          youtubeUrl: topic.youtubeUrl,
          audioUrl: topic.audioUrl,
          pdfUrl: topic.pdfUrl,
          syncCues: topic.syncCues,
        ),
      );
      isChapterStart = false;

      for (final subtopic in topic.subtopics) {
        units.add(
          ReadingUnit(
            unitIndex: units.length,
            chapterId: chapter.id,
            chapterTitle: chapter.title,
            chapterNumber: chapterNumber,
            isChapterStart: false,
            topicNumber: topicNumber,
            topicId: topic.id,
            isSubtopic: true,
            id: subtopic.id,
            title: subtopic.title,
            description: subtopic.description,
            youtubeUrl: subtopic.youtubeUrl,
            audioUrl: subtopic.audioUrl,
            pdfUrl: subtopic.pdfUrl,
            syncCues: subtopic.syncCues,
          ),
        );
      }
    }
  }
  return units;
}

/// A topic and the subtopics filed under it — the sidebar's second level.
///
/// The reader itself pages through the flat [ReadingUnit] list; this grouping
/// exists only so the contents sidebar can show the shape a student recognises
/// from the book, chapter → topic → subtopic, instead of one long list where a
/// subtopic looks like a topic that happens to be indented.
class TopicGroup {
  const TopicGroup({required this.topic, required this.subtopics});

  final ReadingUnit topic;
  final List<ReadingUnit> subtopics;

  bool get hasSubtopics => subtopics.isNotEmpty;

  bool containsUnit(int index) =>
      topic.unitIndex == index ||
      subtopics.any((subtopic) => subtopic.unitIndex == index);

  /// Whether opening this group would reveal any narration — what decides if a
  /// collapsed topic still deserves a headphones hint.
  bool get hasAudioAnywhere =>
      topic.hasAudio || subtopics.any((subtopic) => subtopic.hasAudio);
}

/// A chapter and the span of units it covers — the contents sidebar's model.
class ChapterSummary {
  const ChapterSummary({
    required this.chapterId,
    required this.chapterNumber,
    required this.title,
    required this.units,
    required this.topics,
  });

  final String chapterId;
  final int chapterNumber;
  final String title;

  /// Every unit in the chapter, topics and subtopics alike, in reading order.
  final List<ReadingUnit> units;

  /// The same units as a two-level tree.
  final List<TopicGroup> topics;

  int get unitStart => units.first.unitIndex;
  int get unitEnd => units.last.unitIndex;

  bool containsUnit(int index) => index >= unitStart && index <= unitEnd;
}

/// Splits one chapter's units into topics and the subtopics beneath them.
///
/// [flattenChapters] always emits a topic immediately before its own
/// subtopics, so a single forward pass is enough. A leading subtopic cannot
/// occur, but is given its own group rather than dropped if it ever does.
List<TopicGroup> _groupTopics(List<ReadingUnit> chapterUnits) {
  final groups = <TopicGroup>[];
  for (final unit in chapterUnits) {
    if (!unit.isSubtopic || groups.isEmpty) {
      groups.add(TopicGroup(topic: unit, subtopics: <ReadingUnit>[]));
    } else {
      groups.last.subtopics.add(unit);
    }
  }
  return groups;
}

List<ChapterSummary> buildChapterSummaries(List<ReadingUnit> units) {
  // Units arrive in reading order, so a chapter's units are always contiguous.
  final byChapter = <List<ReadingUnit>>[];
  for (final unit in units) {
    if (byChapter.isNotEmpty &&
        byChapter.last.first.chapterId == unit.chapterId) {
      byChapter.last.add(unit);
    } else {
      byChapter.add(<ReadingUnit>[unit]);
    }
  }

  return [
    for (final chapterUnits in byChapter)
      ChapterSummary(
        chapterId: chapterUnits.first.chapterId,
        chapterNumber: chapterUnits.first.chapterNumber,
        title: chapterUnits.first.chapterTitle,
        units: chapterUnits,
        topics: _groupTopics(chapterUnits),
      ),
  ];
}

/// Locates the unit a saved progress row points at. Prefers the exact topic,
/// then falls back to the chapter's first unit, matching the website's rule.
int resolveResumeIndex(
  List<ReadingUnit> units, {
  String? topicId,
  String? chapterId,
}) {
  if (topicId != null) {
    final byTopic =
        units.indexWhere((u) => !u.isSubtopic && u.id == topicId);
    if (byTopic >= 0) return byTopic;
  }
  if (chapterId != null) {
    final byChapter = units.indexWhere((u) => u.chapterId == chapterId);
    if (byChapter >= 0) return byChapter;
  }
  return -1;
}
